extensions/HyperVExtension/src/HyperVExtension/Services/VMGalleryService.cs

```diff
@@ -13,83 +13,106 @@ namespace HyperVExtension.Services;
 /// <summary>
 /// Service that performs operations specific to the VM gallery.
 /// </summary>
-public sealed class VMGalleryService : IVMGalleryService
+public sealed class VMGalleryService : IVMGalleryService, IDisposable
 {
     private readonly ILogger _log = Log.ForContext("SourceContext", ComponentName);
 
     private const string ComponentName = "VMGalleryService";
 
-    private readonly IDownloaderService _downloaderService;
+    private readonly IDownloaderService _downloaderService;
+
+    private readonly SemaphoreSlim _gallerySemaphore = new(1, 1);
 
     private readonly JsonSerializerOptions _jsonOptions = new()
     {
         PropertyNameCaseInsensitive = true,
         PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
         TypeInfoResolver = JsonSourceGenerationContext.Default,
         AllowTrailingCommas = true,
-    };
+    };
+
+    private bool _disposed;
+
+    public const ulong Windows11ImageExtractedSize = 48318382080UL; // 45 GB in bytes

```

Comment:
We actually do use the ZipFile and ZipArchive apis to get the size when extracting the archive file. However thats only after we download the file. In the scenario I'm attempting to fix the archive file is not already on the users computer. E.g The user hasn't downloaded it yet but they have enough space to download it but they don't have enough space to extract that vhdx file within it after its downloaded. Unfortunately, the information in the VM Gallery Json isn't enough for us to know the size of the vhdx file inside the archive file without first downloading the archive file. So we need a way to prevent the user from downloading the archive file if we know they don't have enough space to fully extract it. I created this issue: https://github.com/microsoft/devhome/issues/3714 for this reason so we can update the gallery Json with this information. These hardcoded values are temporary until that is fixed.

But yes to answer your question if the vhdx files in the archives are updated they will change, so thats why for each image I added a bit more than needed. That should cover us for a long time, and hopefully I can get the gallery Json updated in the meantime

E.g 

1. Win 11 vhdx file is ~42 GB after extraction. So I made the required extracted size 45 GB
2. Win 10 vhdx file  is ~9.5 GB after extraction. So I made the space required 15 GB
3. The ubuntu vhd files are around ~6Gb after extraction. So I made the space required 10 GB